diff --git a/obfs4proxy/obfs4proxy.go b/obfs4proxy/obfs4proxy.go
index 1dd00b2..c267858 100644
--- a/obfs4proxy/obfs4proxy.go
+++ b/obfs4proxy/obfs4proxy.go
@@ -27,10 +27,9 @@
 
 // Go language Tor Pluggable Transport suite.  Works only as a managed
 // client/server.
-package main
+package obfs4proxy
 
 import (
-	"flag"
 	"fmt"
 	"io"
 	golog "log"
@@ -85,7 +84,20 @@ func clientSetup() (launched bool, listeners []net.Listener) {
 			continue
 		}
 
-		ln, err := net.Listen("tcp", socksAddr)
+		realSocksAddr := socksAddr
+		if name == "obfs4" {
+			realSocksAddr = "127.0.0.1:47351"
+		} else if name == "meek_lite" {
+			realSocksAddr = "127.0.0.1:47352"
+		} else if name == "obfs2" {
+			realSocksAddr = "127.0.0.1:47353"
+		} else if name == "obfs3" {
+			realSocksAddr = "127.0.0.1:47354"
+		} else if name == "scramblesuit" {
+			realSocksAddr = "127.0.0.1:47355"
+		}
+
+		ln, err := net.Listen("tcp", realSocksAddr)
 		if err != nil {
 			_ = pt.CmethodError(name, err.Error())
 			continue
@@ -304,22 +316,16 @@ func getVersion() string {
 	return fmt.Sprintf("obfs4proxy-%s", obfs4proxyVersion)
 }
 
-func main() {
+func Start(logLevelStr *string, enableLogging *bool, unsafeLogging *bool) {
 	// Initialize the termination state monitor as soon as possible.
 	termMon = newTermMonitor()
 
-	// Handle the command line arguments.
-	_, execName := path.Split(os.Args[0])
-	showVer := flag.Bool("version", false, "Print version and exit")
-	logLevelStr := flag.String("logLevel", "ERROR", "Log level (ERROR/WARN/INFO/DEBUG)")
-	enableLogging := flag.Bool("enableLogging", false, "Log to TOR_PT_STATE_LOCATION/"+obfs4proxyLogFile)
-	unsafeLogging := flag.Bool("unsafeLogging", false, "Disable the address scrubber")
-	flag.Parse()
-
-	if *showVer {
-		fmt.Printf("%s\n", getVersion())
-		os.Exit(0)
+	if *logLevelStr == "" {
+		*logLevelStr = "ERROR"
 	}
+
+	execName := "obfs4"
+
 	if err := log.SetLogLevel(*logLevelStr); err != nil {
 		golog.Fatalf("[ERROR]: %s - failed to set log level: %s", execName, err)
 	}
@@ -338,8 +344,7 @@ func main() {
 		golog.Fatalf("[ERROR]: %s - failed to initialize logging", execName)
 	}
 	if err = transports.Init(); err != nil {
-		log.Errorf("%s - failed to initialize transports: %s", execName, err)
-		os.Exit(-1)
+		log.Noticef("%s - failed to initialize transports: %s", execName, err)
 	}
 
 	log.Noticef("%s - launched", getVersion())
@@ -379,3 +384,11 @@ func main() {
 	}
 	termMon.wait(true)
 }
+
+func Stop() {
+	log.Noticef("obfs4 Stop: synthesizing SIGINT and SIGTERM")
+
+	termMon.sigChan <- syscall.SIGINT
+
+	termMon.sigChan <- syscall.SIGTERM
+}
diff --git a/obfs4proxy/proxy_http.go b/obfs4proxy/proxy_http.go
index 1adadf8..7ff608d 100644
--- a/obfs4proxy/proxy_http.go
+++ b/obfs4proxy/proxy_http.go
@@ -25,7 +25,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"bufio"
diff --git a/obfs4proxy/proxy_socks4.go b/obfs4proxy/proxy_socks4.go
index ac8be9e..8922e15 100644
--- a/obfs4proxy/proxy_socks4.go
+++ b/obfs4proxy/proxy_socks4.go
@@ -31,7 +31,7 @@
  * license that can be found in the LICENSE file.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"errors"
diff --git a/obfs4proxy/pt_extras.go b/obfs4proxy/pt_extras.go
index 18bc2df..766811e 100644
--- a/obfs4proxy/pt_extras.go
+++ b/obfs4proxy/pt_extras.go
@@ -25,7 +25,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"errors"
diff --git a/obfs4proxy/termmon.go b/obfs4proxy/termmon.go
index 59304c9..53ba1e2 100644
--- a/obfs4proxy/termmon.go
+++ b/obfs4proxy/termmon.go
@@ -25,7 +25,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"io"
diff --git a/obfs4proxy/termmon_linux.go b/obfs4proxy/termmon_linux.go
index 926e630..abbadde 100644
--- a/obfs4proxy/termmon_linux.go
+++ b/obfs4proxy/termmon_linux.go
@@ -25,7 +25,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"fmt"
