diff --git a/obfs4proxy/obfs4proxy.go b/obfs4proxy/obfs4proxy.go
index f295926..5a2ce99 100644
--- a/obfs4proxy/obfs4proxy.go
+++ b/obfs4proxy/obfs4proxy.go
@@ -27,10 +27,9 @@
 
 // Go language Tor Pluggable Transport suite.  Works only as a managed
 // client/server.
-package main
+package obfs4proxy
 
 import (
-	"flag"
 	"fmt"
 	"io"
 	golog "log"
@@ -38,6 +37,7 @@ import (
 	"net/url"
 	"os"
 	"path"
+	"strconv"
 	"sync"
 	"syscall"
 
@@ -51,6 +51,7 @@ import (
 
 const (
 	obfs4proxyVersion = "0.0.14"
+ 	Obfs4proxyVersion = obfs4proxyVersion
 	obfs4proxyLogFile = "obfs4proxy.log"
 	socksAddr         = "127.0.0.1:0"
 )
@@ -58,7 +59,7 @@ const (
 var stateDir string
 var termMon *termMonitor
 
-func clientSetup() (launched bool, listeners []net.Listener) {
+func clientSetup(meekPort, obfs2Port, obfs3Port, obfs4Port, scramblesuitPort *int) (launched bool, listeners []net.Listener) {
 	ptClientInfo, err := pt.ClientSetup(transports.Transports())
 	if err != nil {
 		golog.Fatal(err)
@@ -85,7 +86,20 @@ func clientSetup() (launched bool, listeners []net.Listener) {
 			continue
 		}
 
-		ln, err := net.Listen("tcp", socksAddr)
+		realSocksAddr := socksAddr
+		if name == "obfs4" {
+			realSocksAddr = net.JoinHostPort("127.0.0.1", strconv.Itoa(*obfs4Port))
+		} else if name == "meek_lite" {
+			realSocksAddr = net.JoinHostPort("127.0.0.1", strconv.Itoa(*meekPort))
+		} else if name == "obfs2" {
+			realSocksAddr = net.JoinHostPort("127.0.0.1", strconv.Itoa(*obfs2Port))
+		} else if name == "obfs3" {
+			realSocksAddr = net.JoinHostPort("127.0.0.1", strconv.Itoa(*obfs3Port))
+		} else if name == "scramblesuit" {
+			realSocksAddr = net.JoinHostPort("127.0.0.1", strconv.Itoa(*scramblesuitPort))
+		}
+
+		ln, err := net.Listen("tcp", realSocksAddr)
 		if err != nil {
 			_ = pt.CmethodError(name, err.Error())
 			continue
@@ -304,22 +318,16 @@ func getVersion() string {
 	return fmt.Sprintf("obfs4proxy-%s", obfs4proxyVersion)
 }
 
-func main() {
+func Start(meekPort, obfs2Port, obfs3Port, obfs4Port, scramblesuitPort *int, logLevelStr *string, enableLogging *bool, unsafeLogging *bool) {
 	// Initialize the termination state monitor as soon as possible.
 	termMon = newTermMonitor()
 
-	// Handle the command line arguments.
-	_, execName := path.Split(os.Args[0])
-	showVer := flag.Bool("version", false, "Print version and exit")
-	logLevelStr := flag.String("logLevel", "ERROR", "Log level (ERROR/WARN/INFO/DEBUG)")
-	enableLogging := flag.Bool("enableLogging", false, "Log to TOR_PT_STATE_LOCATION/"+obfs4proxyLogFile)
-	unsafeLogging := flag.Bool("unsafeLogging", false, "Disable the address scrubber")
-	flag.Parse()
-
-	if *showVer {
-		fmt.Printf("%s\n", getVersion())
-		os.Exit(0)
+	if *logLevelStr == "" {
+		*logLevelStr = "ERROR"
 	}
+
+	execName := "obfs4"
+
 	if err := log.SetLogLevel(*logLevelStr); err != nil {
 		golog.Fatalf("[ERROR]: %s - failed to set log level: %s", execName, err)
 	}
@@ -338,8 +346,7 @@ func main() {
 		golog.Fatalf("[ERROR]: %s - failed to initialize logging", execName)
 	}
 	if err = transports.Init(); err != nil {
-		log.Errorf("%s - failed to initialize transports: %s", execName, err)
-		os.Exit(-1)
+		log.Noticef("%s - failed to initialize transports: %s", execName, err)
 	}
 
 	log.Noticef("%s - launched", getVersion())
@@ -347,7 +354,7 @@ func main() {
 	// Do the managed pluggable transport protocol configuration.
 	if isClient {
 		log.Infof("%s - initializing client transport listeners", execName)
-		launched, ptListeners = clientSetup()
+		launched, ptListeners = clientSetup(meekPort, obfs2Port, obfs3Port, obfs4Port, scramblesuitPort)
 	} else {
 		log.Infof("%s - initializing server transport listeners", execName)
 		launched, ptListeners = serverSetup()
@@ -379,3 +386,11 @@ func main() {
 	}
 	termMon.wait(true)
 }
+
+func Stop() {
+	log.Noticef("obfs4 Stop: synthesizing SIGINT and SIGTERM")
+
+	termMon.sigChan <- syscall.SIGINT
+
+	termMon.sigChan <- syscall.SIGTERM
+}
diff --git a/obfs4proxy/proxy_http.go b/obfs4proxy/proxy_http.go
index 1adadf8..7ff608d 100644
--- a/obfs4proxy/proxy_http.go
+++ b/obfs4proxy/proxy_http.go
@@ -25,7 +25,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"bufio"
diff --git a/obfs4proxy/proxy_socks4.go b/obfs4proxy/proxy_socks4.go
index ac8be9e..8922e15 100644
--- a/obfs4proxy/proxy_socks4.go
+++ b/obfs4proxy/proxy_socks4.go
@@ -31,7 +31,7 @@
  * license that can be found in the LICENSE file.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"errors"
diff --git a/obfs4proxy/pt_extras.go b/obfs4proxy/pt_extras.go
index 18bc2df..766811e 100644
--- a/obfs4proxy/pt_extras.go
+++ b/obfs4proxy/pt_extras.go
@@ -25,7 +25,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"errors"
diff --git a/obfs4proxy/termmon.go b/obfs4proxy/termmon.go
index 59304c9..53ba1e2 100644
--- a/obfs4proxy/termmon.go
+++ b/obfs4proxy/termmon.go
@@ -25,7 +25,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"io"
diff --git a/obfs4proxy/termmon_linux.go b/obfs4proxy/termmon_linux.go
index 926e630..abbadde 100644
--- a/obfs4proxy/termmon_linux.go
+++ b/obfs4proxy/termmon_linux.go
@@ -25,7 +25,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-package main
+package obfs4proxy
 
 import (
 	"fmt"
